gdnative-core/src/object.rs

```diff
@@ -31,121 +71,625 @@ pub unsafe trait GodotObject: crate::private::godot_object::Sealed {
+    /// guarantee that the operation is safe.**
+    #[inline(always)]
+    pub unsafe fn assume_safe_during_if_sane<'a, L: 'a>(self, _lifetime: &'a L) -> Option<&'a T> {
+        self.assume_safe_if_sane::<'a>()
+    }
+
+    /// Manually frees the object.
+    ///
+    /// # Safety
+    ///
+    /// During the call, the underlying object must be valid, and this thread must have
+    /// exclusive access to the object. This pointer must never be used again.
+    #[inline]
+    pub unsafe fn free(self) {
+        self.as_raw().free();
+    }
+}
+
+impl<T: GodotObject + ManuallyManaged + QueueFree> Ptr<T> {
+    /// Queues the object for deallocation in the near future.
+    ///
+    /// # Safety
+    ///
+    /// During the call, the underlying object must be valid, and this thread must have
+    /// exclusive access to the object. This pointer must never be used again.
+    #[inline]
+    pub unsafe fn queue_free(self) {
+        T::godot_queue_free(self.as_ptr())
+    }
+}
+
+impl<T: GodotObject + ManuallyManaged> private::Sealed for Ptr<T> {}
+impl<T: GodotObject + ManuallyManaged> PersistentRef for Ptr<T> {
+    type Target = T;
+
+    #[inline]
+    fn sys(&self) -> *mut sys::godot_object {
+        self.ptr.as_ptr()
+    }
+
+    #[inline]
+    unsafe fn as_raw(&self) -> &RawObject<Self::Target> {
+        RawObject::from_sys_ref_unchecked(self.ptr)
+    }
+
+    #[inline]
+    unsafe fn from_sys(obj: NonNull<sys::godot_object>) -> Self {
+        Ptr {
+            ptr: obj,
+            _marker: PhantomData,
+        }
+    }
+
+    #[inline]
+    unsafe fn move_from_sys(obj: NonNull<sys::godot_object>) -> Self {
+        Self::from_sys(obj)
+    }
+
+    #[inline]
+    unsafe fn init_from_sys(obj: NonNull<sys::godot_object>) -> Self {
+        Self::from_sys(obj)
+    }
 }
 
-/// Initialise the reference count of the object.
+impl<T: GodotObject + ManuallyManaged> Eq for Ptr<T> {}
+impl<T: GodotObject + ManuallyManaged> PartialEq for Ptr<T> {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        self.ptr == other.ptr
+    }
+}
+
+impl<T: GodotObject + ManuallyManaged> Ord for Ptr<T> {
+    #[inline]
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.ptr.cmp(&other.ptr)
+    }
+}
+
+impl<T: GodotObject + ManuallyManaged> PartialOrd for Ptr<T> {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.ptr.partial_cmp(&other.ptr)
+    }
+}
+
+impl<T: GodotObject + ManuallyManaged> Hash for Ptr<T> {
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write_usize(self.ptr.as_ptr() as usize)
+    }
+}
+
+impl<T: GodotObject + ManuallyManaged> Debug for Ptr<T> {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}({:p})", T::class_name(), self.ptr)
+    }
+}
+
+/// A reference to a reference-counted Godot object. This is semantically analogous to a `Rc`
```

Comment:
Yes, technically it is atomic, but that does not make it `Send` because the thing inside is not `Sync`. Technically `Arc<T>` also won't be `Send` if `T` is not `Sync` but that might elude users inexperienced with Rust's thread safety system. I find `Rc` a good compromise here.